primap2/pm2io/_data_reading.py

```diff
@@ -38,6 +40,27 @@
     "NE,NO",
     "NE0",
     "NO, NE",
+    "NO",
+    "C",
+    "NaN",
+    "IE",  # this is a values that should be mapped to 0
+    "IE, NE, NO",
+    "IE,NA,NO",
+    "NA,IE,NO",
+    "NO,NE,IE",
+    "NE, NO",
+]
+# TODO: distiguish between values mapped to NA and values mapped to 0, e.g. IE which
```

Comment:
Well, it's how it's done in primap1. But of course it's much better to do it right now. But I think it needs some discussion. So maybe we leave it as a TODO here and open an issue for it?

Reply:
Based on the function to replace NA values it seems to be very easy to implement once we have decided what to map to 0 and what to NA. So maybe we should implement it right away after all.

Reply:
I think `IE` and `NO` (and their combinations) are quite obviously 0, while `NE` and `nan` are obviously NaN. Everything else is more difficult. `NA` (not applicable) is not necessarily obviously NaN, because `NA` often means something like "does not apply in this category". So, I think `NA` combined with `IE` and/or `NO` needs to be 0 as well. `NA,NE` on the other hand should be NaN. About a lonely `NA` I'm not so sure. Maybe NaN is the safer choice there?

Reply:
Based on these definitions
* NO: not occurring 
* NE: not estimated 
* NA: not applicable
* IE: included elsewhere
* C: confidential
 
I agree with you base mapping and would map NA to NaN: 

* NO: 0
* NE: NaN
* NA: NaN
* IE: 0
* C: NaN (if it's confidential there might be data from other sources we can use)
* -: 0, as it's often used as a substitute for 0 in Excel

For NA, I choose NaN as I would say that not applicable is something based on the reporting guidelines not on the country circumstances. So it might be that there are emissions, but they are not reported under the guidelines. If on the other hand there are just no emissions then this should be the case for all datasets and we're fine with NaN.

Now for the combinations:  

* When IE is included I would use 0
* When NO is included I would use 0 as well
* NA, NE is mapped to NaN as both are mapped to NaN individually

In total we have the following mapping for all the values that occurred so far. We could also use regular expressions to implement the rules above to cover new cases instead of adding to the list all the time.

```
NA_VALUES = {
    "IE": 0,  
    "NE": np.nan,
    "NE0": np.nan,
    "NO": 0,
    "C": np.nan,
    "NaN": np.nan,
    "nan": np.nan,
    "-": 0,
    "NA, NE": np,nan,
    "NA,NE": np.nan,
    "NO,NE": 0, 
    "NO, NE": 0,   
    "NE,NO": 0,
    "NE, NO": 0,
    "IE, NE, NO": 0,
    "IE,NA,NO": 0,
    "NA,IE,NO": 0,
    "NO,NE,IE": 0,
}
```

Reply:
sounds good!

A list is fine, I think, but we could also instead use a function:

```python

_special_codes = {
    'C': np.nan,
    'NaN': np.nan,
    'nan': np.nan,
    '-': 0,
    'NE0': 0
}

def parse_code(code: str) -> float:
    code = code.strip()
    if code in _special_codes:
        return _special_codes[code]

    parts = code.split(',')
    parts = [x.replace('.', '').strip() for x in parts]
    if 'IE' in parts or 'NO' in parts:
        return 0
    if 'NE' in parts or 'NA' in parts:
        return np.nan
    raise ValueError(f'Could not parse code: {code!r}.')

Reply:
The advantage is that stuff like `IE,1` where the "1" is a footnote will also be parsed properly and added dots and spaces are handled generally, so it might work for more things. The disadvantage could be that a function is slower and initially harder to read.

Reply:
The function is nice as it will also catch a lot of cases not on our current list. 

Maybe we could give the user the option to specify a dict for a mapping and we use the function as the default. This would be handy if one just wants to change individual mappings. On the other hand a strict distinction between user defined mappings and internal mappings would also be useful for clarity for the users which means either internal mappings or user defined mappings. (see also the convert_nan discussion below)

Reply:
> Maybe we could give the user the option to specify a dict for a mapping and we use the function as the default.

Sounds good.

I didn't understand your "on the other hand". What is the alternative proposal?

Reply:
First option: user defined rules, after these the function is used for the rest. 
Second option: when the user defines rules only these are used and not the function.

Reply:
I would prefer the second option. If a user wants to change things, they get to define everything themselves.

Reply:
A small addition: I would map just spaces to np.nan because it's likely an empty cell read incorrectly

Reply:
Do you map NE0 to 0 instead of nan on purpose? It's mapped to nan in the list solution but to 0 in the function solution

Reply:
No, probably a mistake because I have no idea what `NE0` means.

Reply:
I've now implemented the changes. Str conversion is controllable from the outside. The function is translated into a dict for all values actually occurring in the DataFrame. I've pushed the changes. More tests and docs to come. The behavior of the internal read_csv functions has changed. They now do not convert any strs to NaN but return the raw csv data.

Reply:
The functions has now been implemented and all data reading functions have been changed to use it (or a user specified conversion). The "read_csv" functions now do not do any string processing but return the raw data. All string processing is done in the "convert_XXX_dataframe_if" functions.